sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -667,9 +667,15 @@ object JdbcUtils extends Logging {
     val getConnection: () => Connection = createConnectionFactory(options)
     val batchSize = options.batchSize
     val isolationLevel = options.isolationLevel
-    df.foreachPartition(iterator => savePartition(
-      getConnection, table, iterator, rddSchema, nullTypes, batchSize, dialect, isolationLevel)
-    )
+    if (options.numPartitions != null && options.numPartitions.toInt != df.rdd.getNumPartitions) {
+      df.repartition(options.numPartitions.toInt).foreachPartition(iterator => savePartition(
```

Comment:
This repeats the `foreachPartition` part twice, when that could be outside the if-else.

I don't know enough to say whether we should add this property. It seems a little funny to set this globally to apply to all dataframes written to a database. I understand the use case is pure SQL, where perhaps that's the only option.